src/lib/best_tip_prover/best_tip_prover.ml

```diff
@@ -0,0 +1,143 @@
+open Core_kernel
+open Coda_base
+open Coda_state
+open Async_kernel
+
+module type Inputs_intf = sig
+  include Transition_frontier.Inputs_intf
+
+  module Transition_frontier :
+    Coda_intf.Transition_frontier_intf
+    with type external_transition_validated := External_transition.Validated.t
+     and type mostly_validated_external_transition :=
+                ( [`Time_received] * Truth.true_t
+                , [`Proof] * Truth.true_t
+                , [`Frontier_dependencies] * Truth.true_t
+                , [`Staged_ledger_diff] * Truth.false_t )
+                External_transition.Validation.with_transition
+     and type transaction_snark_scan_state := Staged_ledger.Scan_state.t
+     and type staged_ledger_diff := Staged_ledger_diff.t
+     and type staged_ledger := Staged_ledger.t
+     and type verifier := Verifier.t
+end
+
+module Make (Inputs : Inputs_intf) :
+  Coda_intf.Best_tip_prover_intf
+  with type transition_frontier := Inputs.Transition_frontier.t
+   and type external_transition := Inputs.External_transition.t
+   and type external_transition_with_initial_validation :=
+              Inputs.External_transition.with_initial_validation
+   and type verifier := Inputs.Verifier.t = struct
+  open Inputs
+
+  module Merkle_list = Merkle_list.Make (struct
+    type value = External_transition.Validated.t
+
+    type context = Transition_frontier.t
+
+    type proof_elem = State_body_hash.t
+
+    type hash = State_hash.t [@@deriving eq]
+
+    let to_proof_elem external_transition =
+      external_transition |> External_transition.Validated.protocol_state
+      |> Protocol_state.body |> Protocol_state.Body.hash
+
+    let get_previous ~context transition =
+      let parent_hash =
+        transition |> External_transition.Validated.protocol_state
+        |> Protocol_state.previous_state_hash
+      in
+      let open Option.Let_syntax in
+      let%map breadcrumb = Transition_frontier.find context parent_hash in
+      With_hash.data
+      @@ Transition_frontier.Breadcrumb.transition_with_hash breadcrumb
+
+    let hash acc body_hash =
+      Protocol_state.hash_abstract ~hash_body:Fn.id
+        {previous_state_hash= acc; body= body_hash}
+  end)
+
+  let prove ~logger frontier =
+    let open Option.Let_syntax in
+    let%map () =
```

Comment:
You can still provide a proof if the length to the tip is < max_length, but you have at least max_length - best_tip_length transitions in your frontier's root history.